### ISSUE TITLE
error calling combine function

### ISSUE THREAD

**Issue by simon-tarr** [Azure/doAzureParallel#151]
Hello, 

I had a computer malfunction this week and had to re-install Windows/R/Packages etc. Since this mishap (I was running doAzureParallel v0.4.1), I have installed v0.5.0 and get an error on task completion.

This is the error:

```
Job Summary: 
Id: job2017xxxxx
Job Preparation Status: Package(s) being installed
Waiting for tasks to complete. . .
  |================================================================================================================| 100%
error calling combine function:
<simpleError: obj$state$numResults <= obj$state$numValues is not TRUE>
Number of errors: 0
```
To clarify, I do get an object returned but it's "corrupted" (the actual issue is hard to explain but take my word for it - the returned map is absolutely nothing like the output prior to the problem!).

**EDIT -** **I'll try and explain. I have chunked my original dataset (n=10760) so that each core in my pool gets an approximately even split (so 10760/128=85). The returned object only appears to have the first 127 results; in other words, only the first chunk or so. My output is actually a geographic map so the first 85 pixels in the raster image have 'expected' values (i.e. non-zero values) whereas the rest of the landscape returns 0. I've therefore concluded that the other chunks  are not being merged, for whatever reason, even though it worked absolutely fine even just a few days ago.**

I am installing a package from GitHub and I have specified this in cluster.json.
The pool creation process appears to work without issue:

```
Booting compute nodes. . . 
  |================================================================================================================| 100%
Your cluster has been registered.
Dedicated Node Count: 0
Low Priority Node Count: 1
```

I have carried out the following troubleshooting:

1. Re-installed version 0.4.1; end up with the same problem
2. Have re-created both cluster.json and credentials.json files and triple-checked these details are correct
3. Tried on two independent computers
4. Triple-checked that all resource files I'm calling upon are present and correct. I have double checked all URLs and everything is as expected.

The log file for one of the nodes can be found here: https://1drv.ms/t/s!AlF85N0khZNl9pMzUPEjjm8GDlfk-w

The stderr.txt file for one of the nodes can be found here: https://1drv.ms/t/s!AlF85N0khZNl9pM0BX-iXBsiAPQZlw

Nothing has changed with my R code since this error appears; it seems to have occurred only after installing the latest version of doAzureParallel on a clean install of Windows.

Many thanks.

**Comment by brnleehng** [comment]
Hi Simon,

I was able to repro the error.

In the new version of doAzureParallel, we assumed that the tasks were an exact even split. Therefore the last task 85 would have some results but many of the results would have NULL values and this would create a corrupted object.. 

setChunkSize(3)
foreach(i = 1:10, .options.azure = list(chunkSize = 3)) %dopar% { i }
Job Summary: 
Id: job2017xxxxx
Waiting for tasks to complete. . .
  |================================================================================================================| 100%
error calling combine function:
<simpleError: obj$state$numResults <= obj$state$numValues is not TRUE>
Number of errors: 0

In this case, task 4 would have '10' in its result, but the combine function was expecting task 4 to also '11' and '12'.

I'm still working on 1, not sure why 0.4.1 wouldn't work anymore. Is rAzureBatch version latest version? 

Thanks!
Brian

**Comment by simon-tarr** [comment]
Hello Brian,

Thanks for the reply. I'm relieved you've been able to repro the error. I can confirm that I'm currently running doAzureParallel 0.5.0 and rAzureBatch 0.5.1.

I'm also a bit confused about why 0.4.1 isn't working. Perhaps I've been running an old version of doAzureParallel but a more recent version of batch? I will try rolling both of these back and will see how this goes.

Thanks again.
Simon

**Comment by simon-tarr** [comment]
I have just tried running doAzureParallel versions 0.4.1 and 0.4.3 with versions 0.3.0 and 0.4.0 of rAzureBatch (as found here: https://github.com/Azure/rAzureBatch/releases). However, the tasks cannot complete and I get the error:
```

Job Summary: 
Id: job20171021110724
Job Preparation Status: Package(s) being installedError in rAzureBatch::getJobPreparationStatus(jobId, content = "parsed",  : 
  unused arguments (filter = filter, select = select)
```

**Comment by brnleehng** [comment]
- Is there some reproducible code that I can try out?
- Are you setting the .errorhandling to 'stop' or 'pass'

**Comment by simon-tarr** [comment]
Hello, creating a reproducible example will be tricky - it's using a GitHub repository and requires some software to be installed via the command line in cluster.json. I would also need to provide you with a .csv file containing the input data. Providing you're happy with me doing that, I can get something together for you?

With regards to error handling; it has been set to 'stop'. I have tried with 'pass' and 'remove' and still no luck. In fact, I think I've tried nearly all combination of arguments in my quest to get things working again but I've had no luck.

Thanks for your continued assistance.

**Comment by simon-tarr** [comment]
A quick update for this afternoon; it's working as expected. It appears that the R package ncdf4 is at fault when specified in .packages. Previous to v0.5.0, I installed netcdf via the command line (to install the netcdf libraries) in cluster.json. I also needed to call the R package 'ncdf4' within .packages as well (or the loop would fail). Since v0.5.0 it seems that installing netcdf via command line and then calling it in .packages causes the issue (to be more specific, the issue where data is returned but it looks garbled and out of order). If I drop the R package from '.packages' everything appears to work as expected.

In addition, .errorhandling needs to be changed from 'stop' to 'pass'.

I'm not sure what has changed since these point releases with regards to the way packages and errors are handled but there we have it.

(To clarify I'm running doAzureParallel v0.5.0 and rAzureBatch 0.5.2)

**Comment by brnleehng** [comment]
Good to hear, you were able to reproducible your results. 

I'll take a look on why the package re-installation is failing for 'ncdf4'. It seems to be a bug, .packages should check if the package is installed, if it's not available then it will install.  

The errorhandling 'pass' has the same behavior as v0.4.0. If a task does not produce a result file, it will return NA values for that specific chunk.

**Comment by simon-tarr** [comment]
The interesting thing is, the model doesn't appear to fail at any point. Changing between 'pass' and 'stop' just seems to change the order the model list is returned, which is very strange. When set to 'stop' it looks as though the merge task is putting the original list back in any order. When it's sent to 'pass', it returns the results in the correct order and I get the map I expect.

**Comment by brnleehng** [comment]
@simon-tarr, 

This PR (#223) sorts them in order regardless of errorhandling setting now.

Thanks,
Brian
